shared/feature/feeding/src/main/java/com/epmedu/animeal/feeding/domain/usecase/GetFeedingPointByPriorityUseCase.kt

```diff
@@ -0,0 +1,53 @@
+package com.epmedu.animeal.feeding.domain.usecase
+
+import com.epmedu.animeal.feeding.domain.model.FeedingPoint
+import com.epmedu.animeal.feeding.domain.model.enum.AnimalState
+import com.epmedu.animeal.feeding.presentation.model.FeedingPointModel
+import com.epmedu.animeal.feeding.presentation.model.MapLocation
+import com.epmedu.animeal.feeding.presentation.model.MapLocation.Companion.toPoint
+import com.epmedu.animeal.foundation.tabs.model.AnimalType
+import com.epmedu.animeal.geolocation.location.distanceInKmTo
+import com.epmedu.animeal.geolocation.location.isNearTo
+
+class GetFeedingPointByPriorityUseCase(
+    val getAllFeedingPointsUseCase: GetAllFeedingPointsUseCase
+) {
+
+    /*
+     * Filters iterable of FeedingPointModel by priority defined at EPMEDU-594:
+     * - 1st: Status = Red, is favorite and is near to user location
+     * - 2nd: Status = Red, is favorite
+     * - 3rd: Status = Red, is not favorite but is near to user location
+     */
+    suspend operator fun invoke(
+        type: AnimalType,
+        userLocation: MapLocation,
+        onFeedingPointFound: (FeedingPointModel) -> Unit,
+    ) {
+        getAllFeedingPointsUseCase(type = type).collect { domainFeedingPoints ->
+            val readyToFeed = domainFeedingPoints
+                .filter { it.animalStatus == AnimalState.RED }
+                .sortByNearestToLocation(userLocation)
+            if (readyToFeed.isNotEmpty()) {
+                val feedPointFound = readyToFeed
+                    .firstOrNull {
+                        it.isFavourite && it.location.toPoint().isNearTo(userLocation.toPoint())
```

Comment:
Why we need those logic? wouldn't it be easier to sort the list by multiple predicates, according to priority?

Reply:
Makes sense from the readability point of view. However a predicate for each priority would iterate completely the whole set of feeding points (internally `findOrNull` and `find` is a `for` cycle with an `if()`) which could increase the amount of processing.

Reply:
Didn't catch your point, talking about those
Varargs predicates
https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.comparisons/compare-by.html